### ISSUE TITLE
Duplicate entry for key 'glpi_groups_tickets.unicity'

### ISSUE THREAD

**Issue by fralla2** [glpi-project/glpi#8823]
**Describe the bug**

Not sure since when this problem occurs but sometime the addAdditionalActors action occurs in double for the same actor group and trigger an error in sql-errors.log file

From what I can see in my error log today, it happened once with the automatic ruleticket assignation at the ticket creation (2021-03-11 10:45) and happened a second time when someone tried to add a group actor manually to the ticket some time later that day (2021-03-11 11:44)

**To reproduce**

Create a ticket and have it assigned to a group from ruleticket or from manually adding a group actor to the ticket

**Logs**

```
[2021-03-11 10:45:02] glpisqllog.ERROR: DBmysql::query() in /var/www/html/glpi-9.5.3-bugfixes/inc/dbmysql.class.php line 309
  *** MySQL query error:
  SQL: INSERT INTO `glpi_groups_tickets` (`type`, `groups_id`, `tickets_id`) VALUES ('2', '215', '74453')
  Error: Duplicate entry '74453-2-215' for key 'glpi_groups_tickets.unicity'
  Backtrace :
  inc/dbmysql.class.php:1148                         
  inc/commondbtm.class.php:602                       DBmysql->insert()
  inc/commondbtm.class.php:1142                      CommonDBTM->addToDB()
  inc/commonitilobject.class.php:2141                CommonDBTM->add()
  inc/commonitilobject.class.php:2043                CommonITILObject->addAdditionalActors()
  inc/ticket.class.php:1914                          CommonITILObject->post_addItem()
  inc/commondbtm.class.php:1143                      Ticket->post_addItem()
  front/tracking.injector.php:62                     CommonDBTM->add()
  {"user":"47@glpi06"} 
[2021-03-11 11:48:38] glpisqllog.ERROR: DBmysql::query() in /var/www/html/glpi-9.5.3-bugfixes/inc/dbmysql.class.php line 309
  *** MySQL query error:
  SQL: INSERT INTO `glpi_groups_tickets` (`type`, `groups_id`, `tickets_id`) VALUES ('2', '187', '74453')
  Error: Duplicate entry '74453-2-187' for key 'glpi_groups_tickets.unicity'
  Backtrace :
  inc/dbmysql.class.php:1148                         
  inc/commondbtm.class.php:602                       DBmysql->insert()
  inc/commondbtm.class.php:1142                      CommonDBTM->addToDB()
  inc/commonitilobject.class.php:2141                CommonDBTM->add()
  inc/commonitilobject.class.php:1065                CommonITILObject->addAdditionalActors()
  inc/ticket.class.php:1169                          CommonITILObject->prepareInputForUpdate()
  inc/commondbtm.class.php:1440                      Ticket->prepareInputForUpdate()
  front/ticket.form.php:72                           CommonDBTM->update()
  {"user":"838@glpi06"} 

```

**Screenshots**

![image](https://user-images.githubusercontent.com/38400135/110834915-1d9f7380-826c-11eb-876a-647d954fc9cc.png)

**Your GLPI setup (you can find it in Setup > General menu, System tab)**

```

[code]   GLPI 9.5.3 (/glpi => /var/www/html/glpi-9.5.3-bugfixes) Installation mode: TARBALL
--

Operating system: Linux glpi06 3.10.0-1127.19.1.el7.x86_64 #1 SMP Tue Aug 25 17:23:54 UTC 2020 x86_64 PHP 7.3.24 apache2handler (Core, PDO, Phar, Reflection, SPL, SimpleXML, Zend OPcache, apache2handler, apcu, bcmath, bz2, 	calendar, ctype, curl, date, dom, exif, fileinfo, filter, ftp, gd, gettext, hash, iconv, imap, intl, json, ldap, libxml, 	mbstring, mysqli, mysqlnd, openssl, pcre, pdo_mysql, pdo_sqlite, posix, session, shmop, soap, sockets, sodium, sqlite3, 	standard, sysvmsg, sysvsem, sysvshm, tidy, tokenizer, wddx, xml, xmlreader, xmlrpc, xmlwriter, xsl, zip, zlib) Setup: max_execution_time="30" memory_limit="256M" post_max_size="100M" safe_mode="" session.save_handler="files" 	upload_max_filesize="100M"  Software: Apache/2.4.6 (CentOS) OpenSSL/1.0.2k-fips PHP/7.3.24 () 	Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/88.0.4324.190 Safari/537.36 Server Software: Source distribution 	Server Version: 8.0.21 	Server SQL Mode:  	Parameters: REDACTED 	Host info: REDACTED via TCP/IP 	 PHP version is at least 7.2.0 - Perfect! Sessions support is available - Perfect! Allocated memory > 64 Mio - Perfect! mysqli extension is installed ctype extension is installed fileinfo extension is installed json extension is installed mbstring extension is installed iconv extension is installed zlib extension is installed curl extension is installed gd extension is installed simplexml extension is installed intl extension is installed ldap extension is installed apcu extension is installed Zend OPcache extension is installed xmlrpc extension is installed CAS extension is installed exif extension is installed zip extension is installed bz2 extension is installed sodium extension is installed Database version seems correct (8.0.21) - Perfect! Timezones seems loaded in database The log file has been created successfully. Write access to /var/www/glpiData/config/ has been validated. Write access to /var/www/glpiData/files has been validated. Write access to /var/www/glpiData/files/_dumps has been validated. Write access to /var/www/glpiData/files/_sessions has been validated. Write access to /var/www/glpiData/files/_cron has been validated. Write access to /var/www/glpiData/files/_graphs has been validated. Write access to /var/www/glpiData/files/_lock has been validated. Write access to /var/www/glpiData/files/_plugins has been validated. Write access to /var/www/glpiData/files/_tmp has been validated. Write access to /var/www/glpiData/files/_cache has been validated. Write access to /var/www/glpiData/files/_rss has been validated. Write access to /var/www/glpiData/files/_uploads has been validated. Write access to /var/www/glpiData/files/_pictures has been validated. Write access to /var/www/html/glpi-9.5.3-bugfixes/marketplace has been validated. For security reasons, SELinux mode should be Enforcing.

GLPI_ROOT: /var/www/html/glpi-9.5.3-bugfixes GLPI_CONFIG_DIR: /var/www/glpiData/config/ GLPI_VAR_DIR: /var/www/glpiData/files GLPI_LOG_DIR: /var/log/glpi GLPI_MARKETPLACE_DIR: /var/www/html/glpi-9.5.3-bugfixes/marketplace GLPI_USE_CSRF_CHECK: 1 GLPI_CSRF_EXPIRES: 7200 GLPI_CSRF_MAX_TOKENS: 100 GLPI_USE_IDOR_CHECK: 1 GLPI_IDOR_EXPIRES: 7200 GLPI_ALLOW_IFRAME_IN_RICH_TEXT:  GLPI_TELEMETRY_URI: https://telemetry.glpi-project.org GLPI_INSTALL_MODE: TARBALL GLPI_NETWORK_MAIL: glpi@teclib.com GLPI_NETWORK_SERVICES: https://services.glpi-network.com GLPI_MARKETPLACE_PRERELEASES:  GLPI_USER_AGENT_EXTRA_COMMENTS:  GLPI_AJAX_DASHBOARD: 1 GLPI_CALDAV_IMPORT_STATE: 0 GLPI_DEMO_MODE: 0 GLPI_FORCE_EMPTY_SQL_MODE: 1 GLPI_DOC_DIR: /var/www/glpiData/files GLPI_CACHE_DIR: /var/www/glpiData/files/_cache GLPI_CRON_DIR: /var/www/glpiData/files/_cron GLPI_DUMP_DIR: /var/www/glpiData/files/_dumps GLPI_GRAPH_DIR: /var/www/glpiData/files/_graphs GLPI_LOCAL_I18N_DIR: /var/www/glpiData/files/_locales GLPI_LOCK_DIR: /var/www/glpiData/files/_lock GLPI_PICTURE_DIR: /var/www/glpiData/files/_pictures GLPI_PLUGIN_DOC_DIR: /var/www/glpiData/files/_plugins GLPI_RSS_DIR: /var/www/glpiData/files/_rss GLPI_SESSION_DIR: /var/www/glpiData/files/_sessions GLPI_TMP_DIR: /var/www/glpiData/files/_tmp GLPI_UPLOAD_DIR: /var/www/glpiData/files/_uploads GLPI_NETWORK_REGISTRATION_API_URL: https://services.glpi-network.com/api/registration/ GLPI_MARKETPLACE_PLUGINS_API_URI: https://services.glpi-network.com/api/glpi-plugins/ GLPI_I18N_DIR: /var/www/html/glpi-9.5.3-bugfixes/locales GLPI_VERSION: 9.5.3 GLPI_SCHEMA_VERSION: 9.5.4 GLPI_MIN_PHP: 7.2.0 GLPI_YEAR: 2021

htmlawed/htmlawed version 1.2.5 in (/var/www/html/glpi-9.5.3-bugfixes/vendor/htmlawed/htmlawed) phpmailer/phpmailer version 6.1.6 in (/var/www/html/glpi-9.5.3-bugfixes/vendor/phpmailer/phpmailer/src) simplepie/simplepie version 1.5.6 in (/var/www/html/glpi-9.5.3-bugfixes/vendor/simplepie/simplepie/library) tecnickcom/tcpdf version 6.3.5 in (/var/www/html/glpi-9.5.3-bugfixes/vendor/tecnickcom/tcpdf) michelf/php-markdown in (/var/www/html/glpi-9.5.3-bugfixes/vendor/michelf/php-markdown/Michelf) true/punycode in (/var/www/html/glpi-9.5.3-bugfixes/vendor/true/punycode/src) iamcal/lib_autolink in (/var/www/html/glpi-9.5.3-bugfixes/vendor/iamcal/lib_autolink) sabre/dav in (/var/www/html/glpi-9.5.3-bugfixes/vendor/sabre/dav/lib/DAV) sabre/http in (/var/www/html/glpi-9.5.3-bugfixes/vendor/sabre/http/lib) sabre/uri in (/var/www/html/glpi-9.5.3-bugfixes/vendor/sabre/uri/lib) sabre/vobject in (/var/www/html/glpi-9.5.3-bugfixes/vendor/sabre/vobject/lib) laminas/laminas-cache in (/var/www/html/glpi-9.5.3-bugfixes/vendor/laminas/laminas-cache/src) laminas/laminas-i18n in (/var/www/html/glpi-9.5.3-bugfixes/vendor/laminas/laminas-i18n/src) laminas/laminas-serializer in (/var/www/html/glpi-9.5.3-bugfixes/vendor/laminas/laminas-serializer/src) monolog/monolog in (/var/www/html/glpi-9.5.3-bugfixes/vendor/monolog/monolog/src/Monolog) sebastian/diff in (/var/www/html/glpi-9.5.3-bugfixes/vendor/sebastian/diff/src) elvanto/litemoji in (/var/www/html/glpi-9.5.3-bugfixes/vendor/elvanto/litemoji/src) symfony/console in (/var/www/html/glpi-9.5.3-bugfixes/vendor/symfony/console) scssphp/scssphp in (/var/www/html/glpi-9.5.3-bugfixes/vendor/scssphp/scssphp/src) laminas/laminas-mail in (/var/www/html/glpi-9.5.3-bugfixes/vendor/laminas/laminas-mail/src/Protocol) laminas/laminas-mime in (/var/www/html/glpi-9.5.3-bugfixes/vendor/laminas/laminas-mime/src) rlanvin/php-rrule in (/var/www/html/glpi-9.5.3-bugfixes/vendor/rlanvin/php-rrule/src) blueimp/jquery-file-upload in (/var/www/html/glpi-9.5.3-bugfixes/vendor/blueimp/jquery-file-upload/server/php) ramsey/uuid in (/var/www/html/glpi-9.5.3-bugfixes/vendor/ramsey/uuid/src) psr/log in (/var/www/html/glpi-9.5.3-bugfixes/vendor/psr/log/Psr/Log) psr/simple-cache in (/var/www/html/glpi-9.5.3-bugfixes/vendor/psr/simple-cache/src) mexitek/phpcolors in (/var/www/html/glpi-9.5.3-bugfixes/vendor/mexitek/phpcolors/src/Mexitek/PHPColors) guzzlehttp/guzzle in (/var/www/html/glpi-9.5.3-bugfixes/vendor/guzzlehttp/guzzle/src) guzzlehttp/psr7 in (/var/www/html/glpi-9.5.3-bugfixes/vendor/guzzlehttp/psr7/src) wapmorgan/unified-archive in (/var/www/html/glpi-9.5.3-bugfixes/vendor/wapmorgan/unified-archive/src) paragonie/sodium_compat in (/var/www/html/glpi-9.5.3-bugfixes/vendor/paragonie/sodium_compat/src) phpCas version 1.3.8 in (/usr/share/pear)

Server: 'REDACTED', Port: '389', BaseDN: REDACTED', Connection filter: 		'REDACTED', 		RootDN: 'REDACTED', Use TLS: none

Not active

Way of sending emails: SMTP (anonymous@localhost)

Name: 'REDACTED' Active: Yes 	Server: '{outlook.office365.com/imap/ssl}' Login: 'REDACTED' Password: Yes Name: 'REDACTED' Active: Yes 	Server: '{outlook.office365.com/imap/ssl}' Login: 'REDACTED' Password: Yes Name: 'REDACTED' Active: No 	Server: '{outlook.office365.com/imap/ssl}' Login: 'REDACTED' Password: Yes Name: 'REDACTED' Active: Yes 	Server: '{outlook.office365.com/imap/ssl}' Login: 'REDACTED' Password: Yes Name: 'REDACTED' Active: Yes 	Server: '{outlook.office365.com/imap/ssl}' Login: 'REDACTED' Password: Yes Name: 'REDACTED' Active: Yes 	Server: '{outlook.office365.com/imap/ssl}' Login: 'REDACTED' Password: Yes Name: 'REDACTED' Active: Yes 	Server: '{outlook.office365.com/imap/ssl}' Login: 'REDACTED' Password: Yes

advancedplanning     Name: advancedplanning               Version: 0.2.0      State: Enabled 	treeview             Name: Arborescence                   Version: 1.9.0      State: Enabled 	fields               Name: Champs supplémentaires         Version: 1.12.1     State: Enabled 	datainjection        Name: Data Injection                 Version: 2.9.0      State: Enabled 	formcreator          Name: Form Creator                   Version: 2.11.1     State: Enabled 	fusioninventory      Name: FusionInventory                Version: 9.5.0+1.0  State: Enabled 	genericobject        Name: Gestion d'objets               Version: 2.10.0     State: Enabled 	order                Name: Gestion des commandes          Version: 2.6.0      State: Enabled 	manufacturersimports Name: Imports fabricants             Version: 2.3.1      State: Enabled 	taskdrop             Name: TaskDrop                       Version: 1.2.0      State: Enabled
[/code]


```

**Comment by fralla2** [comment]
**After future investigation :** 

- At (2021-03-11 10:45), the requester created the new ticket and saved it
- At (2021-03-11 10:45:02), GLPI processed all rules automatically and because all conditions respected the conditions of rule 289 shown bellow, the system tried to add the group 215 **for the first time** but was already there before by the ITIL category that affect "Group in charge of the hardware" to that group

![image](https://user-images.githubusercontent.com/38400135/110852602-90b2e500-8280-11eb-9231-ea154ca92a12.png)


- At (2021-03-11 11:44), the technician manually assigned the group 187 and saved the ticket.
- At (2021-03-11 11:48), the technician changed the ticket category to "Service des technologies (43)" and saved the ticket
- At (2021-03-11 11:48:38), GLPI processed all rules automatically and because all conditions respected the conditions of rule 297 shown bellow, the system tried to add the group 187 **a second time** and triggered the error

![image](https://user-images.githubusercontent.com/38400135/110851441-0ae26a00-827f-11eb-886c-9fecae5dbf6f.png)




Now 3 questions : 

1 - For the first error, I think it come from the ITIL Category for this ticket maybe that adds the group 215 automatically before processing ticket rules???
2 - For the 2nd error, soult a check be added somewhere to prevent trying to add the same group twice or should I just ignore the sql-errors ???
3 - Shouldn't the pattern for "rulecriteria" be the id of the group when the criteria relate to a group, instead of the name ???

**Comment by cedric-anne** [comment]
Hi,

Thanks for the analysis.

The `CommonITILObject::addAdditionalActors()` method should verify if an actor already exists before adding it. Indeed, there are many sources of additionnal actors (plugins, category assignment, rules, ...), and it would be difficult, in all of them, to do this kind of checks.

For now, SQL error is not blocking, so it has no functionnal impact, but it should still be fixed.

Regards

**Comment by cedric-anne** [comment]
You can ignore this warning, but I reopened this issue as it still need to be fixed.

**Comment by cconard96** [comment]
Fixed via #10826
